// @zerobytes/object-model-js/dist/FieldTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.ComplexTypes = exports.FieldType = void 0;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var FieldType = function FieldType(complexType, Type) {
  _classCallCheck(this, FieldType);

  this.complexType = complexType;
  this.Type = Type;
};

exports.FieldType = FieldType;
var ComplexTypes = {
  ShapedAs: 'ShapedAs',
  ArrayOf: 'ArrayOf',
  InstanceOf: 'InstanceOf',
  IdOf: 'IdOf'
};
exports.ComplexTypes = ComplexTypes;
var _default = {
  String: 'string',
  Integer: 'integer',
  Float: 'float',
  Array: 'array',
  Datetime: 'datetime',
  Date: 'date',
  Time: 'time',
  Boolean: 'boolean',
  Object: 'object',
  ShapedAs: function ShapedAs(Type) {
    return new FieldType(ComplexTypes.ShapedAs, Type);
  },
  ArrayOf: function ArrayOf(Type) {
    return new FieldType(ComplexTypes.ArrayOf, Type);
  },
  InstanceOf: function InstanceOf(Type) {
    return new FieldType(ComplexTypes.InstanceOf, Type);
  },
  IdOf: function IdOf(ModelBase) {
    return new FieldType(ComplexTypes.IdOf, ModelBase);
  }
};
exports["default"] = _default;
